### DISCUSSION TITLE
Targets woudn't run due to error "run more than one pipeline on the same data store"

### DISCUSSION THREAD

**Discussion by dipterix** [ropensci/targets#1237] [Help]
I upgraded `targets` to the latest version and it suddenly gives me this error

> Process ID 56279 is already running a {targets} pipeline with the shared folder as the local data store for data and metadata files. Please do not attempt to run more than one pipeline on the same data store because it will mangle thos important local files. Before trying again, check that process 56279 is really a {targets} pipeline and not a false positive, then terminate it manually. In case of a false positive, remove file shared/meta/process and try again.

To replicate, run one `tar_make(name = name1)`, then wait for this one to finish, then in another R session, run `tar_make()`.

This is a bug because the first process has already finished when the second make process begins.

**Comment by wlandau** [comment]
@dipterix, maybe you called `tar_make()` with `callr_function = NULL)` at some point? If so, https://github.com/ropensci/targets/commit/0a8e07fdbe034b7fb1d3c360d0a984002fb08164 should fix most occurrences of this issue. False positives can still happen if you run `tar_make(callr_function = NULL)` in a different R process and keep that process running, or if the time stamp resolution on your machine is imprecise (like on Windows). If timestamp imprecision is the cause, you can wait 1-2 seconds between calls to `tar_make()` and that should solve the issue. 

If none of those things apply or help, please tell me more about the machine you are running and how you are calling `tar_make()` (and an exact reprex would help, just so we're perfectly clear).

**Comment by dipterix** [comment]
I did call with `tar_make(callr_function = NULL)`! Thanks for pointing out. I was not sure how I could debug it. Can I ask two questions before you release the new version.

1. Will `callr_function = NULL` always or most likely to be supported in the future?
2. As a temporary fix, is it legit to call `targets::tar_destroy("process", ask = FALSE)` to reset the state once I know the pipeline is finished. Should I be worried about any side effects?

Really appreciate it : )

**Reply by wlandau** [reply]
`callr_function = NULL` will always be supported. However, it should never be used for anything serious: it causes the pipeline to run in an unclean environment in which various targets invalidate unpredictably. The only valid uses for `callr_function = NULL` are:

1. For [interactive debugging](https://books.ropensci.org/targets/debugging.html#pause-the-pipeline-with-the-debug-option) to to reproduce and diagnose a known error.
2. If you are building an extension package like `tarchetypes` and you are running are unit tests. In that case, `callr_function = NULL` makes the tests run faster and increases code coverage, and you are safe because the pipelines are not serious and are run inside the temporary isolate environment provided by `testthat::test_that()`. 

> As a temporary fix, is it legit to call targets::tar_destroy("process", ask = FALSE) to reset the state once I know the pipeline is finished. Should I be worried about any side effects?

`targets::tar_destroy("process", ask = FALSE)` should be fine as long as you are not actively using `tar_process()` for anything important (this is extremely uncommon).

**Reply by dipterix** [reply]
Right, `callr_function` is faster when running in shiny app. I try not to use `NULL`, but sometimes have to sacrifice for better user experience.

Thanks for the advice : )
